Adding the ability to configure profanity plugin

mansona opened the pull request:
Howdy folks 👋 

This is essentially an extension of my work on https://github.com/retextjs/retext-profanities/pull/17 that is attempting to allow you to configure the level of "sureness" that a word is profane. 

This is also a "speculative" PR on how I think we might be able to expose this level of config for each of the plugins. Please let me know if it's along the right lines and I'll make sure that it makes sense on all of the exposed APIs 👍 or if you think there is a better way I am happy to move this PR in that direction

wooorm commented:
Hey nice work!

* This still needs docs and cli support
* Just an idea: why not support `sureness` top-level, does it have to be in `profanities`?

mansona commented:
> * This still needs docs and cli support

🎉 I didn't want to add those until I was sure I wasn't a million miles off with the implementation 😂 

> * Just an idea: why not support `sureness` top-level, does it have to be in `profanities`?

So this is related to our discussion on the other PR. "sureness" in the context of retext-profanities kinda makes sense (even though we weren't 100% happy with it) but in the context of `alex` it makes no sense 😞 "sureness about what". We could make it top-level with `profanitySureness` because then it at least makes sense but I quite liked the idea that you had this top-level object that was essentially passed directly into the respective plugin. That way if we add any more configs to retext-profanities it won't require any code change in alex.

Thoughs?

wooorm commented:
@mansona Sorry for the slow reply. Happy holidays and stuff!

> 🎉 I didn't want to add those until I was sure I wasn't a million miles off with the implementation 😂

No it looks good! 👍

> So this is related to our discussion on the other PR. "sureness" in the context of retext-profanities kinda makes sense (even though we weren't 100% happy with it) but in the context of alex it makes no sense 😞 "sureness about what". We could make it top-level with profanitySureness because then it at least makes sense but I quite liked the idea that you had this top-level object that was essentially passed directly into the respective plugin. That way if we add any more configs to retext-profanities it won't require any code change in alex.

Ah, yeah. Makes sense. But allow `profanities` and passing it to `retext-profanities` shows the internals under alex, and we haven’t done that before. Maybe `profanitySureness` instead?

wooorm commented on index.js:
```diff
@@ -36,13 +40,21 @@ function core(value, processor) {
 // Alex.
-function alex(value, allow) {
+function alex(value, config) {
+  let allow
```
```suggestion
  var allow
```

Rest of the codebase uses var, which is fine for now!

wooorm commented on index.js:
```diff
@@ -53,12 +65,12 @@ function htmlParse(value, allow) {
     unified()
       .use(html)
-      .use(rehype2retext, text)
+      .use(rehype2retext, makeText())
```
Should `config` be passed here too, so that HTML supports profanitySureness as well?

wooorm commented on test/api.js:
```diff
@@ -33,6 +33,25 @@ test('alex()', function(t) {
 })
 
+test('alex() with an allow array', function(t) {
+  let {messages} = alex(
```
Could you use `var messages = alex(...).messages` here (and on L45)? Rest of the codebase uses ES5.

mansona commented:
@wooorm I updated everything as you asked 🎉 

One thing to note: I made sure that the API for configuration on text and html matched exactly the functionality of the normal markdown API, including the "array fallback" style execution that you suggested to remove from the documentation. This means that they now support something that is not documented and was never documented, just for the sake of consistency 😂 

Also, as you suggested I removed the documentation for the "allow array" fallback configuration and now we are only documenting the object style of config. I didn't change the implementation (or the tests) so we can still release this as a minor version 👍

wooorm commented:
Released! Thanks @mansona!

mansona commented:
yay! thanks a million @wooorm 🎉
